synapse/storage/databases/main/deviceinbox.py

```diff
@@ -653,41 +653,67 @@ async def _remove_deleted_devices_from_device_inbox(
 
         def _remove_deleted_devices_from_device_inbox_txn(
             txn: LoggingTransaction,
-        ) -> int:
-            """stream_id is not unique
-            we need to use an inclusive `stream_id >= ?` clause,
-            since we might not have deleted all dead device messages for the stream_id
-            returned from the previous query
-
-            Then delete only rows matching the `(user_id, device_id, stream_id)` tuple,
+        ) -> Tuple[int, bool]:
+            """We delete only rows matching the `(user_id, device_id, stream_id)` tuple,
             to avoid problems of deleting a large number of rows all at once
             due to a single device having lots of device messages.
             """
 
-            last_stream_id = progress.get("stream_id", 0)
-
-            sql = """
-                SELECT device_id, user_id, stream_id
-                FROM device_inbox
-                WHERE
-                    stream_id >= ?
-                    AND (device_id, user_id) NOT IN (
-                        SELECT device_id, user_id FROM devices
+            if "max_stream_id" in progress:
+                max_stream_id = progress["max_stream_id"]
+            else:
+                txn.execute("SELECT max(stream_id) FROM device_inbox")
+                # There's a type mismatch here between how we want to type the row and
+                # what fetchone says it returns, but we silence it because we know that
+                # res can't be None.
+                res: Tuple[Optional[int]] = txn.fetchone()  # type: ignore[assignment]
+                if res[0] is None:
+                    return 0, True
+                else:
+                    max_stream_id = res[0]
+
+            start = progress.get("stream_id", 0)
+            stop = start + batch_size
+
+            if self.database_engine.supports_returning:
+                # If the database engine supports the RETURNING clause, use it and do
+                # everything in one go.
+                sql = """
+                    DELETE FROM device_inbox
+                    WHERE
+                        stream_id >= ? AND stream_id < ?
+                        AND (device_id, user_id) NOT IN (
+                            SELECT device_id, user_id FROM devices
+                        )
+                    RETURNING device_id, user_id, stream_id
```

Comment:
we don't need any of this data. all we need is the value of `stop` (so we know where to pick up next time), and we have that without RETURNING.

Reply:
Right, so we don't care about adding more data to the progress?

I was trying to stick to what this comment says:

https://github.com/matrix-org/synapse/blob/9279e5dca090895a38a995574b473fbffb4815c7/synapse/storage/databases/main/deviceinbox.py#L791-L794

Reply:
I don't really know what that comment is meant to mean! I think it might have been a misunderstanding from the original author.

The only reason to store data in the progress table is so that we can read it next time and pick up where we left off. We're not using `user_id` or `device_id`, so there's no point writing them imho.

Reply:
(assuming the answer to this question is no, I'm changing this)